### PR TITLE
[RetroPlayer] Add basic resolution changing

### PR DESCRIPTION
**Pull request by GTechAlpha**
## Description
Add basic resolution changing.

## Motivation and Context
RetroPlayer does not render after resolution change.

## How Has This Been Tested?
PCSX, GenesisPlus, Nestopia. Example: FFVII on PCSX changes resolution every time the in-game menu is accessed.

## Types of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 1]
```diff
@@ -177,7 +183,10 @@ void CRPRenderManager::FrameMove()
 
     if (m_state == RENDER_STATE::CONFIGURING)
     {
+      UpdateResolution();
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: Currently this function is dead code. Its purpose is borrowed from VideoPlayer for adjusting the screen's refresh rate to match the video. Is this a setting that makes sense for games?

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 2]
```diff
@@ -140,20 +150,16 @@ void CRPRenderManager::AddFrame(const uint8_t* data, size_t size, unsigned int w
     {
       std::vector<uint8_t> cachedFrame = std::move(m_cachedFrame);
 
-      if (!m_bHasCachedFrame)
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: removing this check skips an important operation that elides a memory copy. is there a reason for this change? if unrelated to resolution changing it should go in a separate commit.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.h [hunk 3]
```diff
@@ -175,6 +177,7 @@ namespace RETRO
     bool m_bHasCachedFrame = false;
     std::set<std::string> m_failedShaderPresets;
     bool m_bTriggerUpdateResolution = false;
+    std::atomic<bool> m_bFlush;
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.h]: If my c++11 is correct, this can be initialized in the header like `std::atomic<bool> m_bFlush{};`

File: xbmc/cores/RetroPlayer/buffers/BaseRenderBufferPool.cpp [hunk 4]
```diff
@@ -73,16 +73,20 @@ bool CBaseRenderBufferPool::Configure(AVPixelFormat format, unsigned int width,
 
 IRenderBuffer *CBaseRenderBufferPool::GetBuffer(size_t size)
 {
-  if (!m_bConfigured)
+  if (!m_bConfigured) {
+    CLog::Log(LOGDEBUG, "RetroPlayer[RENDER]: buffer pool is not configured");
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/buffers/BaseRenderBufferPool.cpp]: Please place unrelated changes in a separate commit

File: xbmc/cores/RetroPlayer/buffers/BaseRenderBufferPool.cpp [hunk 5]
```diff
@@ -73,16 +73,20 @@ bool CBaseRenderBufferPool::Configure(AVPixelFormat format, unsigned int width,
...
     return nullptr;
+  }
 
   // If frame size is unknown, set it now
   if (m_frameSize == 0)
     m_frameSize = size;
 
   // Changing sizes is not implemented
-  if (m_frameSize != size)
+  if (m_frameSize != size) {
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/buffers/BaseRenderBufferPool.cpp]: Coding style is { on new line. This has the added benefit of lowering the number of deleted lines in a pull request (an important measure)

**Comment by garbear** [comment]
In general this is a good approach. I only found one flaw, which I sent you in https://github.com/GTechAlpha/xbmc/pull/1. Once this gets cleaned up it's good to go.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 6]
```diff
@@ -452,35 +485,37 @@ IRenderBuffer *CRPRenderManager::GetRenderBuffer(IRenderBufferPool *bufferPool)
...
 {
+  if (m_bFlush || m_state != RENDER_STATE::CONFIGURED)
+    return;
+
   CSingleLock lock(m_bufferMutex);
 
   if (!HasRenderBuffer(bufferPool) && m_bHasCachedFrame)
   {
     std::vector<uint8_t> cachedFrame = std::move(m_cachedFrame);
-    if (!cachedFrame.empty())
```
[the remark this answers is not in the document]
Comment (GTechAlpha) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: This is a duplicate check. If m_bHasCachedFrame is true, then there is a frame.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 2 above)
Comment (GTechAlpha) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: As it is asynchronous, the cached frame could be updated? The memcopy will only happen when the emulator sends frames during pause.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: If the check was duplicate, then we could drop `m_bHasCachedFrame`. When writing this I discovered an edge case that caused me to store the cached frame's state externally in the bool. I made the mistake of not documenting why, but I don't think it's 100% safe to drop the second check.

[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: I remember! Instead of explaining here I'll add documentation to the source.

[the remark this answers is not in the document]
Comment (GTechAlpha) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: Re-added

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 7]
```diff
@@ -177,7 +182,13 @@ void CRPRenderManager::FrameMove()
 
     if (m_state == RENDER_STATE::CONFIGURING)
     {
+      CLog::Log(LOGDEBUG, "RetroPlayer[RENDER]: Configuring renderer(s)");
+
+      for (auto &renderer : m_renderers)
+        renderer->Configure(m_format, m_width, m_height);
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: Renderers are created on demand, and configured on creation, so this shouldn't be necessary. Is `m_renderers` even nonempty at this point?

Comment (GTechAlpha) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: This is critical. It is for re-configuring existing renderers on resolution change. I will add a comment to clarify.

Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp]: I see this now. Thanks.

**Comment by garbear** [comment]
I see one more change that's needed:

```c++
std::map<AVPixelFormat, SwsContext*> m_scalers;
```
m_scalers needs to be indexed by format, width and height instead of just format.

**Comment by GTechAlpha** [comment]
This should be good to go, AFAIK.

**Comment by garbear** [comment]
What about this? we could split configuration and reconfiguration: https://github.com/garbear/xbmc/commit/rp-video

Switch to Fullscreen msg should only be sent once. I imagine your amlogic patch doesn't need to be called multiple times either. "re"configuration of the renderers, by definition, doesn't need to happen in the first configuration.

**Comment by garbear** [comment]
Also this: https://github.com/garbear/xbmc/commit/rp-buffers

Buffers need to be protected by the buffer mutex.

**Comment by garbear** [comment]
Also this: https://github.com/garbear/xbmc/commit/rp-buffers

Buffers need to be protected by the buffer mutex.

**Comment by GTechAlpha** [comment]
The buffer is protected by m_bFlush plus this being on the render thread.

**Comment by garbear** [comment]
The game thread can access the buffers.

Also I looked into the `m_scalers` comment above and no changes are needed for this.

**Comment by garbear** [comment]
I see that m_bFlush protects the buffers. Still, should we be defensive?

**Comment by GTechAlpha** [comment]
> What about this? we could split configuration and reconfiguration: rp-vide
> 
> Switch to Fullscreen msg should only be sent once. I imagine your amlogic patch doesn't need to be called multiple times either. "re"configuration of the renderers, by definition, doesn't need to happen in the first configuration. So having separate configuration and reconfiguration states makes sense.

Yes, the Amlogic code only needs to be called once. I think would be a good addition.

**Comment by garbear** [comment]
K, Configure() needs a way to know which state to set

**Comment by GTechAlpha** [comment]
I'll add when I get a chance.

**Comment by GTechAlpha** [comment]
Anything else on this one?

**Comment by garbear** [comment]
Almost :) In the buffer lock commit, `m_bHasCachedFrame` and `m_cachedFrame` should be reset within the lock. Also I'll send you a PR with a reconfig tweak that I think improves the code.

**Comment by garbear** [comment]
@GTechAlpha: https://github.com/GTechAlpha/xbmc/pull/4

**Comment by garbear** [comment]
squash into two commits and i'll merge!

**Comment by GTechAlpha** [comment]
:thumbsup:
